RationaleMCP/0035/Readme.md

```diff
@@ -0,0 +1,123 @@
+* Users can even write their own translation, when having a template *.pot-File for the library. 
+
+Use case for your proposal that show how it is applied. 
+* POT-File of current MSL 4.0 library --> this translation template is provided by ESI and filled for the Hydraulics Pump ... 
+(*add pictures*)
+
+*For each use case state how it could be implemented by current Modelica at best and why the current Modelica does not suffice for this application.*
+
+Proposed changes in Specification: 
+* Add section 13.6 into "external ressources" which describes following:
+* Where shall the translation file be stored (Subfolder Language in the Resources Folder; proposal see below in tool implementation [step 3](#3.-file-storage))
+* How does a GNU gettext translation template for a library look like and which kind of descriptive texts shall be included (proposal in tool implementation [step 1](#1.-read-out-a-translation-template-file))?
+* How does a GNU gettext translation of a library look like (proposal in tool implementation [step 2](#2.-translation-of-the-template-file-into-a-language-file))?
+* Which strings should be translated:
+
+The following Modelica constructs shall be translated: 
+* Description / comment
+* choices annotation
+* missingInnerMessage, obsolete, unassignedMessage annotations
+* Dialog.[group|tab] annotations
+* Dialog.[load|save]Selector.[caption|filter] annotations
+* Documentation.[info|revisions] annotations
+* Text in diagram
+
+The precise updated text of the specification is part of this branch/pull-request: No.
+
+# Backwards Compatibility
+The proposal is backwards compatible. All multilingual text changes can be applied without effect at existing Modelica code. 
+After the introduction of translation files for libraries, this files need to be adjusted with every library change which adds new commented parts or changes existing comments.
+
+# Tool Implementation
+
+## Experience with Prototype
+### Implementation steps
+#### 1. Read out a translation template file
+Create at first a translation template file \*.pot for a Modelica library. Means a (tool dependent) readout of all needed text strings into a *LibraryName*.pot-File. It contains: 
+- a header which concludes all necessary information
+![Header of a *.pot file](POT_Header.png)
+
+- for all original strings entries with following structure: 
+```
+    msgctxt "Hydraulics.Basics.TankOverflow"
+    msgid "Volume Flow Port A"
+    msgstr ""
+```
+
+Where *msgctxt* contains the ident of the Modelica Class and *msgid* contains the textstring which is situated in this name space and shall be translated. Please regard that if a *msgid* string (in this example "Volume Flow Port A") is given more than once in a namespace, all occurrences are translated with the same translation!
+
+It shall be mandatory to use for *msgctxt* the full ident of a class:
```

Comment:
I remember discussing this at length at a meeting in 2018.  From the [notes of that discussion](https://github.com/modelica/ModelicaSpecification/issues/302#issuecomment-435594115):
> Full context (lexical name) vs. just English text as names; conclusion seems to be use lexical name.
>
> What to translate:
> …
>
> General conclusion: fully qualified name+string as key - skip string if not needed.

As I remember it, the meaning of these notes is that one should be able to use a fully qualified path to a component as a more precise alternative to just giving the name of the class and the string to be translated.  For example, consider:
```
model M
  record R
    Real x "Hello";
  end R;
  R r(x "Goodbye");
end M;
```
It would be good to clarify which of the candidate contexts below that should be allowed:
```
# Candidate contexts: "M.R.x", "M.R", "M"
msgctxt "M.R.x"
msgid "Hello"
msgstr ""

# Candidate contexts: "M.r.x", "M.r", "M"
msgctxt "M.r.x"
msgid "Goodbye"
msgstr ""
```

Reply:
In this proposal, only the full name of the class which forms the namespace where the comment belongs to is expected. In the example it means
```
# Candidate contexts: "M.R.x", "M.R", "M" only  "M.R" expected
msgctxt "M.R"
msgid "Hello"
msgstr ""

# Candidate contexts: "M.r.x", "M.r", "M" only "M" expected
msgctxt "M"
msgid "Goodbye"
msgstr ""
```
This simplifies the handling. Also, usually a class forms a semantical context in which same words may have the same translation in other languages. There is no need to specify the location in more detail here.